src/Shared/E2ETesting/WaitAssert.cs

```diff
@@ -73,19 +76,48 @@ private static void WaitAssertCore(IWebDriver driver, Action assertion, TimeSpan
             }
             catch (WebDriverTimeoutException)
             {
+                // At this point at least one test failed, so we mark the test as failed. Any assertions after this one
+                // will fail faster. There's a small race condition here between checking the value for TestRunFailed
+                // above and setting it here, but nothing bad can come out of it. Worst case scenario, one or more
+                // tests running concurrently might use the DefaultTimeout in their current assertion, which is fine.
+                TestRunFailed = true;
+
+                var fileId = $"{Guid.NewGuid():N}.png";
+                var screenShotPath = Path.Combine(Path.GetFullPath(E2ETestOptions.Instance.ScreenShotsPath), fileId);
                 var errors = driver.GetBrowserLogs(LogLevel.Severe);
-                if (errors.Count > 0)
-                {
-                    throw new BrowserAssertFailedException(errors, lastException);
-                }
-                else if (lastException != null)
+
+                if (driver is ITakesScreenshot takesScreenshot && E2ETestOptions.Instance.ScreenShotsPath != null)
                 {
-                    ExceptionDispatchInfo.Capture(lastException).Throw();
+                    try
+                    {
+                        if (!Directory.Exists(E2ETestOptions.Instance.ScreenShotsPath))
+                        {
+                            Directory.CreateDirectory(E2ETestOptions.Instance.ScreenShotsPath);
+                        }
+
+                        var screenShot = takesScreenshot.GetScreenshot();
+                        screenShot.SaveAsFile(screenShotPath);
+                    }
+                    catch (Exception)
+                    {
```

Comment:
Could we log something here? If we failed to write to disk for some reason, it would be good to know why.

Reply:
I added a Console.WriteLine statement here, but I doubt it'll be captured. Sadly there's nothing we can do here.